Updates for newer ES6 features.

RMcNeely opened the pull request:
Previously we did not need to handle many ES6 features correctly so there are a few things that were not checked for initially. This update can now: 

- Parse arrow functions correctly
- Accommodate default arguments in functions
- Removes 'external' references in JSDoc comment when documenting Web APIs

In addition there are now several examples that can be easily tested and stepped thru when using VSCode's node debugger. This makes understanding what is happening a lot easier. Each of these examples has a node script so that they can easily be run without linking the project to another.

RMcNeely committed:
Now parses memebers that are arrow functions correctly.
New test for arrow functions.
Changed variable names for clarity and added whitespace to tests.

RMcNeely committed:
Merge branch 'production' of github.com:ChartIQ/dts-generator into development

RMcNeely committed:
Created some small examples to watch the output locally in development and testing.

RMcNeely committed:
Redo simple class example to not use class fields

RMcNeely committed:
Fixes for default arguments in functions.
Adds semicolons to examples so they are parsed correctly.

RMcNeely committed:
Strip out any "external:" references for documented external Web APIs

normunds-ciq commented on src/code-analysis/comment-collector.js:
```diff
@@ -129,7 +129,8 @@ function applyMemberRoles(members, data) {
 
-    if (member.definition.includes('function')) {
+    // (\(.*\)\s*=>)|\b\s*\(\s*\)|(function)
+    if (/(\(.*\)\s*\s*{)|(\(.*\)\s*\s*=>)/.test(member.definition)) {
```
Why \s* defined 2x?

Why not leave function as one option and simplify to
/function|\(.*\)\s*=>/

RMcNeely replied:
I was trying to target the parens of a function but I think this is probably too repetitive.

RMcNeely replied:
Can you think of any reason not to use this regex? 
```
/function|=>/
```
Originally I thought I needed to capture the arguments and parens when I was working on this but that's not actually the case. I think this can probably suffice.

normunds-ciq replied:
That makes it very readable.  I can not think of a way it would give false positive for what it is intended for.

Commit:
Add getDefinition function that allows extraction of multiline function signatures

Prettier formatting in source code has introduced multiline function signatures that needs to be accomodated.

Commit:
Update function and method parmeter extraction and formatting

Commit:
Prevent crash when JSDoc parameter string does not mach expected patterrn

Commit:
Apply type fix

Commit:
Update complex parameter extraction

Commit:
Update parameter wrap minimum width setting

The value change from 60 to 80 is arbitrary but appears to be working better with Prettier formatting setting at 80 char per line

Commit:
Add back processing of arrow function properties

Removal of arrow function from the definition string prevented function detection

Commit:
Prevent if statement matching "isFunction"

Commit:
Revert to using indeOf for name value split

Commit:
Fix property name extraction

Commit:
Remove empty comment lines in typedef

Commit:
Process `@tsdeclaration` tags

Commit:
Move static class members to namespace

Commit:
Update callback definition to type rather than function declaration

Commit:
Update logging

Commit:
Temporary fix for AdvancedInjectable

Commit:
Recognize `@instance` tag to select instance properties

Commit:
Move to namespace only class static methods

Commit:
Add export to automatically created namespace

Commit:
Update ~ to be replaced with . rather than .prototype.

~ in JSDoc refers to inner function and often is used in as type.  In TS it make sense to attach type to namespace object directly.

RMcNeely committed:
Remove .only from test and set GH Actions to run on development branch

Commit:
Add support for destructured arguments

For exampe CIQ.UI.Chart#createChartAndUI

Commit:
Add guard for not defined argument

Commit:
Use arg value when definition name not available

RMcNeely committed:
Support fixing types for union types

RMcNeely committed:
Coerce optional arguments not at the end of a function to be a union type with undefined.
Add test for coercion

RMcNeely committed:
Add support for optional objects

Commit:
Add rest parameter processing

Commit:
Ensure class instance method defined as part of class is not marked as static method

Commit:
Prevent "if (" case being interpreted as class method

Commit:
Process static class functions

Commit:
Fix types in union type for individual items, remove return from class defined constuctor

Commit:
Simplify fixType addressing union types without wrapping ()

Commit:
Include "*" in fixType processing

Commit:
Include in fixType processng types starting with "external:"

RMcNeely committed:
Adds support for customized export tag

RMcNeely committed:
Add pre and post processing hooks to config.

Commit:
Allow @tsdeclaration overwrites for constructors

Commit:
Fix duplicate definition error

Commit:
Do not generate info on issues with deprecated memeber definitions

RMcNeely committed:
Move pre and post hooks into the generate function

RMcNeely committed:
Update test expectations

Commit:
Add configurable option to include @private by default set to false

Commit:
Extend object property declared interface based to default object values

Extension application is managed in configuration

Commit:
Apply `@tsdeclaration` overwrite to class members

For situation where JSDocs can not satisfy TS declaration requirements

Commit:
Place `@tsdeclaration` overriden functions in namespace

Pick up on starting function keyword in addition to the 'public static ...' to determine function placement

Commit:
Remove unnecessary console.log, remove false positive parmeter and argument mismatch on rest parameter

Commit:
Add memberof identification for mismatched parameters in JSDoc and function signature

Commit:
Improve selectability of info messages

TSgen prefix and consistent Parameter start of info message related to parameter mismatch simplifies grep of message type when run in build process

RMcNeely committed:
Bump minor version
